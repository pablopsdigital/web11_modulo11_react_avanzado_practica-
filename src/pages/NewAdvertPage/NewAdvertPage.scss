#new-advert-page {
  margin: 4rem 0;
  .header {
    h2 {
      font-weight: 600;
    }
  }

  .body {
    .columns {
      display: grid;
      grid-auto-columns: 1fr;
      grid-template-columns: 1fr 1fr;
      gap: 5rem;

      .input-container {
        margin-top: 2rem;
        label {
          font-weight: 600;
        }
      }
    }
    .footer {
      display: flex;
      flex-direction: row;
      align-items: center;
      justify-content: space-between;
      margin-top: 3rem;
    }
  }
}
