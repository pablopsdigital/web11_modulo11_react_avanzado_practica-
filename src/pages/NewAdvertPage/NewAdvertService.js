import ApiClient from '../../services/ApiClient';

export const createAdvertisement = (advertisement) => {
  const url = `/api/v1/adverts`;
  return ApiClient.post(url, advertisement);
};
