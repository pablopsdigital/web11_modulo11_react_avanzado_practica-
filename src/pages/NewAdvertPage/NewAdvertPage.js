import Layout from '../../containers/Layout/Layout';
import { useState, useEffect } from 'react';
import Button from '../../components/Button/Button';
import './NewAdvertPage.scss';
import InputFile from '../../components/InputFile/InputFile';
import imageNoPhoto from '../../images/no-image.png';
import { createAdvertisement } from './NewAdvertService';
import { Redirect } from 'react-router-dom';
import SpinnerLoading from '../../components/SpinnerLoading/SpinnerLoading';
import Alert from '../../components/Alert/Alert';
import { getAllTags } from '../../components/FiltersForm/FiltersService';
import { Link } from 'react-router-dom/cjs/react-router-dom.min';

function NewAdvertPage({ ...props }) {
  //Data
  const [isLoading, setIsLoading] = useState(true);
  const [error, setError] = useState(null);

  const resetError = () => {
    setError(null);
  };

  useEffect(() => {
    resetError();
    setIsLoading(false);
  }, []);

  //Name
  const [name, setName] = useState('');
  const handleInputName = (event) => {
    setName(event.target.value);
  };

  //Sale
  const [type, setSale] = useState(null);
  const handleInputSale = (event) => {
    setSale(event.target.value);
    console.log(type);
  };

  //Price
  const [price, setPrice] = useState(0);
  const handleInputPrice = (event) => {
    setPrice(event.target.value);
  };

  //Tags
  const [selectTags, setSelectTags] = useState([]);
  const [tags, setTags] = useState([]);

  useEffect(() => {
    setIsLoading(true);
    getAllTags().then((tags) => setTags(tags));
    setIsLoading(false);
  }, []);

  const handleCheckTag = (event) => {
    var listTags = [...selectTags];
    if (event.target.checked) {
      listTags = [...selectTags, event.target.value];
    } else {
      listTags.splice(selectTags.indexOf(event.target.value), 1);
    }
    setSelectTags(listTags);
  };

  //Photo
  const [photo, setPhoto] = useState();
  const [photoRender, setRenderPhoto] = useState({ imageNoPhoto });
  const handleInputPhoto = (event) => {
    const reader = new FileReader();

    if (event.target.files[0]) {
      setPhoto(event.target.files[0]);
    }

    reader.onload = () => {
      if (reader.readyState === 2) {
        setRenderPhoto({ imageNoPhoto: reader.result });
      }
    };

    if (event.target.files[0]) {
      reader.readAsDataURL(event.target.files[0]);
    }
  };

  //Create and save new advert
  const [createIdAdvertResponse, setIdAdverResponse] = useState('');
  const createdAdvert = async (newAdvertFormData) => {
    try {
      const createdAdvertResponse = await createAdvertisement(newAdvertFormData);
      setIdAdverResponse(createdAdvertResponse.id);
    } catch (error) {
      setError(error);
    }
  };

  //Send form
  const handleFormSubmit = async (event) => {
    event.preventDefault();
    resetError();
    setIsLoading(true);

    if (!selectTags.length) {
      setError({ message: 'Select one tags' });
    } else {
      const newAdvertFormData = new FormData();
      newAdvertFormData.set('name', name);
      newAdvertFormData.set('sale', type);
      newAdvertFormData.set('price', price);
      newAdvertFormData.set('tags', selectTags);
      if (photo) {
        newAdvertFormData.set('photo', photo);
      }

      createdAdvert(newAdvertFormData);
    }
    setIsLoading(false);
  };
  //Redirect;
  if (createIdAdvertResponse) {
    return <Redirect to={`/adverts/${createIdAdvertResponse}`} />;
  }

  return (
    <Layout {...props}>
      <div id="new-advert-page">
        <div className="container">
          <div className="header">
            <h2>NewAdvertPage</h2>
          </div>

          <div className="body">
            <form onSubmit={handleFormSubmit}>
              <div className="columns">
                <div className="colum">
                  <div className="input-container">
                    <label>Name</label>
                    <input
                      type="text"
                      className=""
                      placeholder="name"
                      value={name}
                      onChange={handleInputName}
                      required
                    />
                  </div>

                  <div className="input-container">
                    <label>Price (â‚¬)</label>
                    <input
                      type="number"
                      className=""
                      placeholder="price"
                      value={price}
                      onChange={handleInputPrice}
                      required
                    />
                  </div>

                  <div>
                    <div className="input-container">
                      <label>Type advert: </label>
                      <input
                        name="sale"
                        type="radio"
                        value="true"
                        checked={type === 'true'}
                        onChange={handleInputSale}
                        required
                      />
                      Sale
                      <input
                        name="sale"
                        type="radio"
                        value="false"
                        checked={type === 'false'}
                        onChange={handleInputSale}
                        required
                      />
                      Buy
                    </div>
                  </div>

                  <div className="input-container">
                    <label>Tags: </label>
                    <ul>
                      {tags.map((tag, index) => (
                        <li key={tag}>
                          <input
                            name="type"
                            type="checkbox"
                            onChange={(e) => handleCheckTag(e)}
                            value={tag}
                          />
                          {tag}
                        </li>
                      ))}
                    </ul>
                  </div>
                </div>

                <div className="colum">
                  <div className="input-container">
                    <label>Image: </label>
                    <InputFile
                      onChange={handleInputPhoto}
                      imageNoPhoto={photoRender}
                      validFormats={'image/*'}
                      name={'image-upload'}
                      {...props}
                    />
                  </div>
                </div>
              </div>
              <div className="footer">
                <div>
                  <Link to="/">
                    <Button>Cancel</Button>
                  </Link>
                </div>
                <div>
                  <Button type="submit">Create Advert</Button>
                </div>
              </div>
            </form>

            {isLoading && <SpinnerLoading />}
            {error && <Alert onClick={resetError}>{error.message}</Alert>}
          </div>
        </div>
      </div>
    </Layout>
  );
}

export default NewAdvertPage;
