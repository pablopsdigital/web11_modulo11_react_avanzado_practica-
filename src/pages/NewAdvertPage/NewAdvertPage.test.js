import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import NewAdvertPage from './NewAdvertPage';

describe('NewAdvertPage', () => {
  test('snapshot NewAdvertPage', () => {
    const store = {
      getState: () => ({
        auth: false,
        adverts: {
          loaded: false,
          data: [],
          tags: []
        },
        ui: {
          isLoading: false,
          error: null
        },
        filters: {}
      }),
      subscribe: () => {},
      dispatch: () => {}
    };
    const tree = renderer
      .create(
        <Provider store={store}>
          <BrowserRouter>
            <NewAdvertPage />
          </BrowserRouter>
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
