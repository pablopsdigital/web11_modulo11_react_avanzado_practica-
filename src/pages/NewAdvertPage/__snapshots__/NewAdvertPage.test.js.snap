// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewAdvertPage snapshot NewAdvertPage 1`] = `
<div>
  <nav
    className="navigation"
    id="navbar"
  >
    <div
      className="menu-icon"
    />
    <div
      className="navigation container"
    >
      <div
        className="brand-container"
      >
        <a
          href="/"
          onClick={[Function]}
        >
          <svg
            alt="brand"
            className="brand-img"
          >
            brand.svg
          </svg>
        </a>
      </div>
      <div
        className="nav-options"
      >
        <a
          aria-current={null}
          href="/adverts"
          onClick={[Function]}
        >
          Adverts
        </a>
        <a
          aria-current={null}
          href="/adverts/new"
          onClick={[Function]}
        >
          New Advert
        </a>
        <a
          href="/login"
          onClick={[Function]}
        >
          <button
            className="login-button"
          >
            Login
          </button>
        </a>
      </div>
    </div>
  </nav>
  <div
    className="container"
  >
    <main>
      <section>
        <div
          id="new-advert-page"
        >
          <div
            className="container"
          >
            <div
              className="header"
            >
              <h2>
                NewAdvertPage
              </h2>
            </div>
            <div
              className="body"
            >
              <form
                onSubmit={[Function]}
              >
                <div
                  className="columns"
                >
                  <div
                    className="colum"
                  >
                    <div
                      className="input-container"
                    >
                      <label>
                        Name
                      </label>
                      <input
                        className=""
                        onChange={[Function]}
                        placeholder="name"
                        required={true}
                        type="text"
                        value=""
                      />
                    </div>
                    <div
                      className="input-container"
                    >
                      <label>
                        Price (€)
                      </label>
                      <input
                        className=""
                        onChange={[Function]}
                        placeholder="price"
                        required={true}
                        type="number"
                        value={0}
                      />
                    </div>
                    <div>
                      <div
                        className="input-container"
                      >
                        <label>
                          Type advert: 
                        </label>
                        <input
                          checked={false}
                          name="sale"
                          onChange={[Function]}
                          required={true}
                          type="radio"
                          value="true"
                        />
                        Sale
                        <input
                          checked={false}
                          name="sale"
                          onChange={[Function]}
                          required={true}
                          type="radio"
                          value="false"
                        />
                        Buy
                      </div>
                    </div>
                    <div
                      className="input-container"
                    >
                      <label>
                        Tags: 
                      </label>
                      <ul />
                    </div>
                  </div>
                  <div
                    className="colum"
                  >
                    <div
                      className="input-container"
                    >
                      <label>
                        Image: 
                      </label>
                      <div
                        className="container-input-file"
                      >
                        <h3
                          className="input-file-head"
                        >
                          Add your Image
                        </h3>
                        <div
                          className="img-holder"
                        >
                          <img
                            alt=""
                            className="img"
                            id="img"
                            src="no-image.png"
                          />
                        </div>
                        <input
                          formats="image/*"
                          id="input-file"
                          name="image-upload"
                          onChange={[Function]}
                          type="file"
                        />
                        <div
                          className="input-field-label"
                        >
                          <label
                            className="image-upload"
                            htmlFor="input-file"
                          >
                            Choose your image
                          </label>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  className="footer"
                >
                  <div>
                    <a
                      href="/"
                      onClick={[Function]}
                    >
                      <button>
                        Cancel
                      </button>
                    </a>
                  </div>
                  <div>
                    <button
                      type="submit"
                    >
                      Create Advert
                    </button>
                  </div>
                </div>
              </form>
            </div>
          </div>
        </div>
      </section>
    </main>
  </div>
  <footer
    id="footer"
  >
    <div
      className="container"
    >
      <div
        className="grid"
      >
        <section
          className="brand"
        >
          <div
            className="brand-container"
          >
            <a
              href="/"
              onClick={[Function]}
            >
              <svg
                alt="brand"
                className="brand-img"
              >
                brand.svg
              </svg>
            </a>
          </div>
          <p
            className="brand-tagline"
          >
            © Todos los derechos reservados
          </p>
        </section>
        <section
          className="links1"
        >
          <nav>
            <ul>
              <li>
                <a
                  href="/404.html"
                  onClick={[Function]}
                >
                  About us
                </a>
              </li>
              <li>
                <a
                  href="/404.html"
                  onClick={[Function]}
                >
                  Contact
                </a>
              </li>
              <li>
                <a
                  href="/404.html"
                  onClick={[Function]}
                >
                  NodePop PRO
                </a>
              </li>
              <li>
                <a
                  href="/404.html"
                  onClick={[Function]}
                >
                  Press
                </a>
              </li>
            </ul>
          </nav>
        </section>
        <section
          className="links2"
        >
          <nav>
            <ul>
              <li>
                <a
                  href="/404.html"
                  onClick={[Function]}
                >
                  Carrers
                </a>
              </li>
              <li>
                <a
                  href="/404.html"
                  onClick={[Function]}
                >
                  Support
                </a>
              </li>
              <li>
                <a
                  href="/404.html"
                  onClick={[Function]}
                >
                  Privacy policy
                </a>
              </li>
              <li>
                <a
                  href="/404.html"
                  onClick={[Function]}
                >
                  Cookies
                </a>
              </li>
            </ul>
          </nav>
        </section>
        <section>
          <p>
            Follow us:
          </p>
          <nav>
            <ul
              className="social-icons"
            >
              <li>
                <a
                  href="https://www.facebook.com/keepcoding.training/"
                >
                  <svg
                    className="social-icon"
                  >
                    icon-facebook.svg
                  </svg>
                </a>
              </li>
              <li>
                <a
                  href="https://www.youtube.com/channel/UCz-oGx94gqD1lICJQZGniLA"
                >
                  <svg
                    className="social-icon"
                  >
                    icon-youtube.svg
                  </svg>
                </a>
              </li>
              <li>
                <a
                  href="https://twitter.com/intent/follow?screen_name=KeepCoding_"
                >
                  <svg
                    className="social-icon"
                  >
                    icon-twitter.svg
                  </svg>
                </a>
              </li>
              <li>
                <a
                  href="https://www.instagram.com/keepcoding_es/?hl=es"
                >
                  <svg
                    className="social-icon"
                  >
                    icon-instagram.svg
                  </svg>
                </a>
              </li>
            </ul>
          </nav>
        </section>
      </div>
    </div>
  </footer>
</div>
`;
